src/utils/binarize.ts

```diff
@@ -25,42 +25,25 @@ export function binarize(
 ): void {
   grayscale(srcImageData, dstImageData)
 
-  const { data: dst, width, height } = dstImageData
-
-  if (typeof threshold === 'undefined') {
-    if (isRGBA(dstImageData)) {
-      const size = width * height
-      const dstGrayData = new Uint8Array(size)
-      const rgbaData = dstImageData.data
-
-      for (
-        let rgbaOffset = 0, grayOffset = 0;
-        grayOffset < size;
-        grayOffset += 1, rgbaOffset += 4
-      ) {
-        dstGrayData[grayOffset] = rgbaData[rgbaOffset]
-      }
-
-      threshold = otsu(dstGrayData)
-    } else {
-      threshold = otsu(dstImageData.data)
-    }
-  }
+  const { data: dst } = dstImageData
+  threshold ??= otsu(dstImageData.data, getImageChannelCount(dstImageData))
 
   if (isRGBA(dstImageData)) {
-    const size = dstImageData.data.length
-
-    for (let offset = 0; offset < size; offset += 4) {
-      const pixel = dst[offset] < threshold ? PIXEL_BLACK : PIXEL_WHITE
-      dst[offset] = pixel
-      dst[offset + 1] = pixel
-      dst[offset + 2] = pixel
-      dst[offset + 3] = 0xff
+    const dst32 = new Int32Array(dst.buffer)
+    const RGBA_WHITE =
+      PIXEL_WHITE | (PIXEL_WHITE << 8) | (PIXEL_WHITE << 16) | (0xff << 24)
+    const RGBA_BLACK =
+      PIXEL_BLACK | (PIXEL_BLACK << 8) | (PIXEL_BLACK << 16) | (0xff << 24)
+
```

Comment:
does this change the endianness? I see `0xff` being shifted all the way to the left, while in the previous code it was `dst[offset + 3]` that is set to `0xff`. Maybe I'm confused?

Reply:
That's something I was confused about as well. The docs for `Int32Array` do say that it's platform-dependent as to the byte ordering, so that's something to be aware of here. Here's something that shows how I came to this math:

```ts
> bytes = Uint8ClampedArray.of(1,2,3,0xff)
Uint8ClampedArray(4) [ 1, 2, 3, 255 ]
> int32s = new Int32Array(bytes.buffer)
Int32Array(1) [ -16580095 ]
> 1 | 2 << 8 | 3 << 16 | 0xff << 24
-16580095
```

Also, all the tests pass so I'm reasonably sure this is right.

Reply:
wild. OK!

Reply:
LGTY to approve?

Reply:
oh yeah, duh